### PR TITLE
 Added crimes viewPager and the ability to edit them from there

### PR DESCRIPTION
**Pull request by f7deleon**
Added crimes viewPager and the ability to edit them from there

### Related Trello cards:
https://trello.com/c/HbAolsuE

### UI preview:
7rrxlgrtuc

### CODE REVIEW

File: app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt [hunk 1]
```diff
@@ -21,7 +21,13 @@ class CrimeAdapter(crimes: List<Crime>, private val callback: (Crime) -> Unit) :
...
     override fun getItemCount() = crimes.size
 
-    fun notifyItemChangedByID(index: UUID) = notifyItemChanged(crimes.indexOfFirst { it.id == index })
+    fun notifyItemChangedByID(ids: ArrayList<String>) {
+        ids.asSequence().forEach {
+            notifyItemChanged(crimeIDStringToIndex(it))
+        }
+    }
+
+    fun crimeIDStringToIndex(str :String): Int =crimes.indexOfFirst{ it.id == UUID.fromString(str)}
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt]: detektcheck

File: app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt [hunk 2]
```diff
@@ -21,7 +21,13 @@ class CrimeAdapter(crimes: List<Crime>, private val callback: (Crime) -> Unit) :
 
     override fun getItemCount() = crimes.size
 
-    fun notifyItemChangedByID(index: UUID) = notifyItemChanged(crimes.indexOfFirst { it.id == index })
+    fun notifyItemChangedByID(ids: ArrayList<String>) {
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt]: add `@MainThread`

File: app/src/main/java/com/example/felipe/criminalintent/CrimeActivity.kt [hunk 3]
```diff
@@ -18,7 +17,7 @@ class CrimeActivity : SingleFragmentActivity() {
     private var fragment: CrimeFragment? = null
     override fun initFragments() {
         val crimeId = intent.getSerializableExtra(EXTRA_CRIME_ID) as? UUID
-        fragment = CrimeFragment.newInstance(crimeId?:UUID.randomUUID())
+        fragment = CrimeFragment.newInstance(crimeId ?: UUID.randomUUID(), {})
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimeActivity.kt]: it should be in the default constructor

File: app/src/main/java/com/example/felipe/criminalintent/CrimePageAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+class CrimePageAdapter(crimes: List<Crime>, val onCrimeUpdated: (Crime) -> Unit, fragmentManager: FragmentManager) : FragmentStatePagerAdapter(fragmentManager) {
+    private val crimes = ArrayList<Crime>(crimes)
+
+    override fun getItem(position: Int): Fragment {
+        val crime = crimes.get(position)
+        return CrimeFragment.newInstance(crime.id, onCrimeUpdated)
+    }
+
+    override fun getCount(): Int {
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimePageAdapter.kt]: sl

File: app/src/main/java/com/example/felipe/criminalintent/CrimePagerActivity.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        btnLast.setOnClickListener {
+            crimeViewPager.currentItem = (crimeViewPager.adapter as CrimePageAdapter).count
+        }
+    }
+
+    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
+        return if (keyCode == KeyEvent.KEYCODE_BACK) {
+            val intent = Intent()
+            intent.putStringArrayListExtra(CrimeFragment.EXTRA_CRIME_ID, updatedCrimes)
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimePagerActivity.kt:46]: didn't you add an extension?

File: app/src/main/java/com/example/felipe/criminalintent/CrimeFragment.kt [hunk 6]
```diff
@@ -43,18 +46,21 @@ class CrimeFragment : Fragment() {
...
         with(view) {
-            chkSolved.isChecked = crime?.isSolved?:false
-            chkPoliceRequired.isChecked = crime?.isPoliceRequire?:false
+            chkSolved.isChecked = crime?.isSolved ?: false
+            chkPoliceRequired.isChecked = crime?.isPoliceRequire ?: false
             titleEditText.setText(crime?.title)
             btnCrime.text = crime?.date.toString()
             chkSolved.setOnCheckedChangeListener { _, isChecked ->
                 crime?.isSolved = isChecked
+                onCrimeUpdated.invoke(crime ?: Crime())
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimeFragment.kt]: maybe you can use let here

**Review by mirland — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt (hunk 2 above)
Comment (FerrariBruno) [reply app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt]: Always reference the interfaces (List in this case)

File: app/src/main/res/layout/activity_crime_pager.xml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    >
+
+    <android.support.v4.view.ViewPager
+        xmlns:android="http://schemas.android.com/apk/res/android"
+        android:id="@+id/crimeViewPager"
+        android:layout_width="match_parent"
+        android:layout_height="464dp"
```
Comment (FerrariBruno) [inline app/src/main/res/layout/activity_crime_pager.xml]: Did you test in small screens? Like a 3.7 or 4 inch screen. Maybe you could use the weight atribute. Thoughts @matir91 ?

**Review by FerrariBruno — APPROVED** [review]

File: app/src/main/res/layout/activity_crime_pager.xml (hunk 7 above)
Comment (mirland) [reply app/src/main/res/layout/activity_crime_pager.xml]: From my point of view it should be in a relative, so you will not have this problem

File: app/src/main/res/layout/activity_crime_pager.xml [hunk 8]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    >
+
+    <android.support.v4.view.ViewPager
+        xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (mirland) [inline app/src/main/res/layout/activity_crime_pager.xml]: rm

File: app/detekt-config.yml [hunk 9]
```diff
@@ -30,176 +49,433 @@ console-reports:
...
+    active: false
+    allowedNames: "(_|ignored|expected)"
+  UseDataClass:
+    active: false
+    excludeAnnotatedClasses: ""
+  UtilityClassWithPublicConstructor:
+    active: false
+  WildcardImport:
+    active: true
+    excludeImports: 'java.util.*,kotlinx.android.synthetic.*'
```
Comment (mirland) [inline app/detekt-config.yml]: nl

File: app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt [hunk 10]
```diff
@@ -21,7 +22,14 @@ class CrimeAdapter(crimes: List<Crime>, private val callback: (Crime) -> Unit) :
...
 
-    fun notifyItemChangedByID(index: UUID) = notifyItemChanged(crimes.indexOfFirst { it.id == index })
+    @MainThread
+    fun notifyItemChangedByID(ids: ArrayList<String>) {
+        ids.asSequence().forEach {
+            notifyItemChanged(crimeIDStringToIndex(it))
+        }
+    }
+
+    fun crimeIDStringToIndex(str :String): Int = crimes.indexOfFirst{ it.id == UUID.fromString(str)}
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt]: getCrimeIndex

Comment (mirland) [reply app/src/main/java/com/example/felipe/criminalintent/CrimeAdapter.kt]: private

File: app/src/main/java/com/example/felipe/criminalintent/CrimeActivity.kt [hunk 11]
```diff
@@ -18,7 +17,7 @@ class CrimeActivity : SingleFragmentActivity() {
     private var fragment: CrimeFragment? = null
     override fun initFragments() {
         val crimeId = intent.getSerializableExtra(EXTRA_CRIME_ID) as? UUID
-        fragment = CrimeFragment.newInstance(crimeId?:UUID.randomUUID())
+        fragment = CrimeFragment.newInstance(crimeId , {})
```
Comment (mirland) [inline app/src/main/java/com/example/felipe/criminalintent/CrimeActivity.kt]: remember it

**Review by mirland — APPROVED** [review]

**Pull request merged**